relayer/src/chain/cosmos/gas.rs

```diff
@@ -42,21 +46,46 @@ pub fn mul_ceil(a: u64, f: f64) -> BigInt {
     (f * a).ceil().to_integer()
 }
 
-/// Adjusts the fee based on the configured `gas_adjustment` to prevent out of gas errors.
-/// The actual gas cost, when a transaction is executed, may be slightly higher than the
-/// one returned by the simulation.
-fn adjust_gas_with_simulated_fees(config: &GasConfig, gas_amount: u64) -> u64 {
-    let gas_adjustment = config.gas_adjustment;
-
-    assert!(gas_adjustment <= 1.0);
+/// Multiply `a` with `f` and round the result down to the nearest integer.
+pub fn mul_floor(a: u64, f: f64) -> BigInt {
+    assert!(f.is_finite());
 
-    let (_, digits) = mul_ceil(gas_amount, gas_adjustment).to_u64_digits();
-    assert!(digits.len() == 1);
+    let a = BigInt::from(a);
+    let f = BigRational::from_float(f).expect("f is finite");
+    (f * a).floor().to_integer()
+}
 
-    let adjustment = digits[0];
-    let gas = gas_amount.checked_add(adjustment).unwrap_or(u64::MAX);
+struct AdjustGas {
+    gas_multiplier: f64,
+    max_gas: u64,
+    gas_amount: u64,
+}
 
-    min(gas, config.max_gas)
+/// Adjusts the fee based on the configured `gas_multiplier` to prevent out of gas errors.
+/// The actual gas cost, when a transaction is executed, may be slightly higher than the
+/// one returned by the simulation.
+fn adjust_estimated_gas(
+    AdjustGas {
+        gas_multiplier,
+        max_gas,
+        gas_amount,
+    }: AdjustGas,
+) -> u64 {
+    assert!(gas_multiplier >= 1.0);
+
+    // Multiply the gas estimate by the gas_multiplier option
+    let (_, digits) = mul_floor(gas_amount, gas_multiplier).to_u64_digits();
+
+    let gas = if digits.len() == 1 {
+        // If the result fits in a u64, use that
+        digits[0]
+    } else {
+        // Otherwise, use u64::MAX
+        u64::MAX
+    };
```

Comment:
The case where the `gas_amount` is simulated to 0 will cause to have the computed gas being u64::MAX, and thus the value returned by `adjust_estimated_gas` to be the `max_gas` configured.
Is this expected ? In the case this is expected, should there be a trace warning or error to inform that the simulated gas for the tx was 0 ?

Reply:
Good catch, thanks! Fixed in https://github.com/informalsystems/ibc-rs/pull/2353/commits/b0cbc4fd8ef8500f634448553867b01a265e07d8 and added short-circuiting for when `gas_amount = 0` or `gas_multiplier = 1.0` in [Short-circuit when gas_amount = 0 or gas_multiplier = 1](https://github.com/informalsystems/ibc-rs/pull/2353/commits/341774d1839f574ca193382e9e8903931f3fa966)